File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/state/DemoViewModel.kt
package ai.zetic.demo.posemotion.state

import ai.zetic.demo.posemotion.AppConfig
import ai.zetic.demo.posemotion.pipeline.BenchmarkSnapshot
import ai.zetic.demo.posemotion.pipeline.FrameResult
import ai.zetic.demo.posemotion.pipeline.MelangeModel
import ai.zetic.demo.posemotion.pipeline.PosePipeline
import ai.zetic.demo.posemotion.video.ClipFrameSource
import ai.zetic.demo.posemotion.video.ClipLocator
import ai.zetic.demo.posemotion.video.VideoFrameDecoder
import android.app.Application
import android.graphics.PointF
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import androidx.lifecycle.AndroidViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

data class ModelLoadState(
    val label: String,
    val progress: Float = 0f,
    val loaded: Boolean = false,
    val error: String? = null,
    val optional: Boolean = false,
)

class DemoViewModel(app: Application) : AndroidViewModel(app) {
    enum class Phase { LOADING, MISSING_CLIP, RUNNING }

    private val inferenceThread = HandlerThread("inference").apply { start() }
    private val handler = Handler(inferenceThread.looper)
    private val mainHandler = Handler(Looper.getMainLooper())
    private val pipeline = PosePipeline()
    private var source: ClipFrameSource? = null

    private val _phase = MutableStateFlow(Phase.LOADING)
    val phase = _phase.asStateFlow()

    private val _loadStates = MutableStateFlow(
        listOf(
            ModelLoadState("YOLO26n · detector"),
            ModelLoadState("RTMPose-s · 2D pose"),
            ModelLoadState("MotionBERT-Lite · 3D lift", optional = true),
        )
    )
    val loadStates = _loadStates.asStateFlow()

    private val _mode = MutableStateFlow(ClipFrameSource.Mode.BENCHMARK)
    val mode = _mode.asStateFlow()

    private val _show3D = MutableStateFlow(true)
    val show3D = _show3D.asStateFlow()

    private val _frame = MutableStateFlow<FrameResult?>(null)
    val frame = _frame.asStateFlow()

    private val _stats = MutableStateFlow(BenchmarkSnapshot())
    val stats = _stats.asStateFlow()

    private val _ballTrail = MutableStateFlow<List<PointF>>(emptyList())
    val ballTrail = _ballTrail.asStateFlow()

    private val _availableClips = MutableStateFlow<List<String>>(emptyList())
    val availableClips = _availableClips.asStateFlow()

    private val _selectedClip = MutableStateFlow(AppConfig.CLIP_FILES.first())
    val selectedClip = _selectedClip.asStateFlow()

    val liftAvailable: Boolean get() = pipeline.lift.isLoaded

    private data class Spec(val model: MelangeModel, val name: String, val version: Int)
    private val specs = listOf(
        Spec(pipeline.detector, AppConfig.DETECTOR_NAME, AppConfig.DETECTOR_VERSION),
        Spec(pipeline.pose, AppConfig.POSE_NAME, AppConfig.POSE_VERSION),
        Spec(pipeline.lift, AppConfig.LIFT_NAME, AppConfig.LIFT_VERSION),
    )

    init {
        pipeline.onResult = { result, snapshot ->
            mainHandler.post { apply(result, snapshot) }
        }
        // Sequential loads on the inference thread: SDK native init is thread-affine,
        // so models must be constructed on the thread that runs them.
        handler.post {
            specs.forEachIndexed { index, spec -> loadOne(index, spec) }
            pipeline.warmup()
            mainHandler.post { startIfReady() }
        }
    }

    fun retry(index: Int) {
        updateRow(index) { it.copy(error = null, progress = 0f) }
        handler.post {
            loadOne(index, specs[index])
            pipeline.warmup()
            mainHandler.post { startIfReady() }
        }
    }

    fun toggle3D() {
        _show3D.value = !_show3D.value
    }

    fun setMode(newMode: ClipFrameSource.Mode) {
        if (newMode == _mode.value) return
        _mode.value = newMode
        if (_phase.value != Phase.RUNNING) return
        source?.stop()
        handler.post { pipeline.resetSession() }
        _ballTrail.value = emptyList()
        startSource()
    }

    fun setClip(fileName: String) {
        if (fileName == _selectedClip.value) return
        _selectedClip.value = fileName
        if (_phase.value != Phase.RUNNING) return
        source?.stop()
        handler.post { pipeline.resetSession() }
        _ballTrail.value = emptyList()
        _frame.value = null
        startSource()
    }

    /** Re-checks for clips (missing-clip screen "check again"). */
    fun recheckClip() {
        if (_phase.value != Phase.MISSING_CLIP) return
        val clips = ClipLocator.availableClips(getApplication())
        if (clips.isNotEmpty()) {
            _availableClips.value = clips
            if (_selectedClip.value !in clips) _selectedClip.value = clips.first()
            _phase.value = Phase.RUNNING
            startSource()
        }
    }

    override fun onCleared() {
        source?.stop()
        handler.post { pipeline.deinitAll() }
        inferenceThread.quitSafely()
    }

    // MARK: - Internals

    /** Runs on the inference thread. */
    private fun loadOne(index: Int, spec: Spec) {
        try {
            spec.model.load(getApplication(), spec.name, spec.version) { progress ->
                mainHandler.post { updateRow(index) { it.copy(progress = progress) } }
            }
            mainHandler.post { updateRow(index) { it.copy(loaded = true, error = null) } }
        } catch (t: Throwable) {
            mainHandler.post {
                updateRow(index) { it.copy(error = t.message ?: t.javaClass.simpleName) }
            }
        }
    }

    private fun startIfReady() {
        if (_phase.value != Phase.LOADING) return
        val allRequired = _loadStates.value.all { it.loaded || (it.optional && it.error != null) }
        if (!allRequired) return
        val clips = ClipLocator.availableClips(getApplication())
        _availableClips.value = clips
        if (clips.isEmpty()) {
            _phase.value = Phase.MISSING_CLIP
            return
        }
        if (_selectedClip.value !in clips) _selectedClip.value = clips.first()
        _phase.value = Phase.RUNNING
        startSource()
    }

    private fun startSource() {
        // Fresh ClipSource per start: the decoder closes the asset fd on release.
        val clip = ClipLocator.locate(getApplication(), _selectedClip.value) ?: run {
            _phase.value = Phase.MISSING_CLIP
            return
        }
        val s = ClipFrameSource(VideoFrameDecoder(clip), _mode.value, handler)
        s.onFrame = { frame -> pipeline.process(frame) }
        source = s
        s.start()
    }

    private fun updateRow(index: Int, transform: (ModelLoadState) -> ModelLoadState) {
        _loadStates.value = _loadStates.value.mapIndexed { i, s ->
            if (i == index) transform(s) else s
        }
    }

    private fun apply(result: FrameResult, snapshot: BenchmarkSnapshot) {
        _frame.value = result
        _stats.value = snapshot
        result.ballBox?.let { ball ->
            val trail = _ballTrail.value + PointF(ball.centerX(), ball.centerY())
            _ballTrail.value = trail.takeLast(AppConfig.BALL_TRAIL_LENGTH)
        }
    }
}

File: apps/Pose-Motion/iOS/PoseMotion/State/DemoViewModel.swift
import Combine
import CoreGraphics
import Foundation
import SwiftUI
import ZeticMLange

struct ModelLoadState: Identifiable {
    let id: String            // display name
    var progress: Float = 0
    var loaded = false
    var error: String?
    var optional = false
}

@MainActor
final class DemoViewModel: ObservableObject {
    enum Phase {
        case loading
        case missingClip
        case running
    }

    @Published var phase: Phase = .loading
    @Published var loadStates: [ModelLoadState] = [
        ModelLoadState(id: "YOLO26n · detector"),
        ModelLoadState(id: "RTMPose-s · 2D pose"),
        ModelLoadState(id: "MotionBERT-Lite · 3D lift", optional: true),
    ]
    @Published var mode: ClipFrameSource.Mode = .benchmark
    @Published var show3D = true
    @Published var availableClips: [String] = []
    @Published var selectedClip = AppConfig.clipNames[0]

    // Latest frame state
    @Published var frameImage: CGImage?
    @Published var frameSize: CGSize = .zero
    @Published var personBox: CGRect?
    @Published var ballBox: CGRect?
    @Published var keypoints: [Keypoint2D]?
    @Published var pose3D: [SIMD3<Float>]?
    @Published var ballTrail: [CGPoint] = []
    @Published var stats = BenchmarkSnapshot()

    private let pipeline = PosePipeline()
    private var source: ClipFrameSource?

    var liftAvailable: Bool { pipeline.lift.isLoaded }

    var allRequiredLoaded: Bool {
        loadStates.allSatisfy { $0.loaded || ($0.optional && $0.error != nil) }
    }

    /// Bundled (blue-folder Media/ or bundle root), then Documents for pushing a clip
    /// via Finder/Files without rebuilding.
    /// ("Media", not "Resources" — a top-level Resources/ dir breaks iOS codesign.)
    private func clipURL(for name: String) -> URL? {
        if let url = Bundle.main.url(forResource: name,
                                     withExtension: AppConfig.clipExtension,
                                     subdirectory: "Media") { return url }
        if let url = Bundle.main.url(forResource: name,
                                     withExtension: AppConfig.clipExtension) { return url }
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
        if let url = docs?.appendingPathComponent("\(name).\(AppConfig.clipExtension)"),
           FileManager.default.fileExists(atPath: url.path) { return url }
        return nil
    }

    /// The configured clips that actually resolve on this device, in config order.
    private func refreshAvailableClips() {
        availableClips = AppConfig.clipNames.filter { clipURL(for: $0) != nil }
        if !availableClips.contains(selectedClip), let first = availableClips.first {
            selectedClip = first
        }
    }

    var clipURL: URL? { clipURL(for: selectedClip) }

    init() {
        pipeline.onResult = { [weak self] result, snapshot in
            self?.apply(result, snapshot)
        }
        loadModels()
    }

    func loadModels() {
        load(index: 0, model: pipeline.detector,
             name: AppConfig.detectorName, version: AppConfig.detectorVersion,
             target: AppConfig.detectorTarget, apType: AppConfig.detectorAPType)
        load(index: 1, model: pipeline.pose,
             name: AppConfig.poseName, version: AppConfig.poseVersion)
        load(index: 2, model: pipeline.lift,
             name: AppConfig.liftName, version: AppConfig.liftVersion)
    }

    func retry(index: Int) {
        loadStates[index].error = nil
        loadStates[index].progress = 0
        switch index {
        case 0: load(index: 0, model: pipeline.detector,
                     name: AppConfig.detectorName, version: AppConfig.detectorVersion,
                     target: AppConfig.detectorTarget, apType: AppConfig.detectorAPType)
        case 1: load(index: 1, model: pipeline.pose,
                     name: AppConfig.poseName, version: AppConfig.poseVersion)
        default: load(index: 2, model: pipeline.lift,
                      name: AppConfig.liftName, version: AppConfig.liftVersion)
        }
    }

    func startIfReady() {
        guard allRequiredLoaded, phase == .loading else { return }
        refreshAvailableClips()
        guard let url = clipURL else {
            phase = .missingClip
            return
        }
        phase = .running
        start(url: url)
    }

    /// Re-checks for a clip (missing-clip screen "try again").
    func recheckClip() {
        guard phase == .missingClip else { return }
        refreshAvailableClips()
        if let url = clipURL {
            phase = .running
            start(url: url)
        }
    }

    func setClip(_ name: String) {
        guard name != selectedClip else { return }
        selectedClip = name
        guard phase == .running, let url = clipURL else { return }
        source?.stop()
        pipeline.resetSession()
        ballTrail.removeAll()
        frameImage = nil
        start(url: url)
    }

    func setMode(_ newMode: ClipFrameSource.Mode) {
        guard newMode != mode else { return }
        mode = newMode
        guard phase == .running, let url = clipURL else { return }
        source?.stop()
        pipeline.resetSession()
        ballTrail.removeAll()
        start(url: url)
    }

    private func start(url: URL) {
        let s = ClipFrameSource(url: url, mode: mode, queue: pipeline.queue)
        s.onFrame = { [weak self] pixelBuffer, time in
            self?.pipeline.process(pixelBuffer, at: time)
        }
        source = s
        s.start()
    }

    private func load(index: Int, model: MelangeModel, name: String, version: Int?,
                      target: Target? = nil, apType: APType = .NA) {
        model.load(
            name: name, version: version, target: target, apType: apType,
            onProgress: { [weak self] p in self?.loadStates[index].progress = p },
            completion: { [weak self] result in
                guard let self else { return }
                switch result {
                case .success:
                    self.loadStates[index].loaded = true
                case .failure(let error):
                    self.loadStates[index].error = error.localizedDescription
                }
                self.startIfReady()
            }
        )
    }

    private func apply(_ result: FrameResult, _ snapshot: BenchmarkSnapshot) {
        frameImage = result.image
        frameSize = result.frameSize
        personBox = result.personBox
        ballBox = result.ballBox
        keypoints = result.keypoints
        pose3D = result.pose3D
        stats = snapshot

        if let ball = result.ballBox {
            ballTrail.append(CGPoint(x: ball.midX, y: ball.midY))
            if ballTrail.count > AppConfig.ballTrailLength {
                ballTrail.removeFirst(ballTrail.count - AppConfig.ballTrailLength)
            }
        }
    }
}
